drivers/platform/mbed/mbed_spi.cpp

```diff
@@ -0,0 +1,280 @@
+		/* Free the CSB gpio object */
+		if ((DigitalOut *)(((struct mbed_spi_desc *)(desc->extra))->csb_gpio)) {
+			delete((DigitalOut *)(((struct mbed_spi_desc *)(desc->extra))->csb_gpio));
+		}
+	}
+
+	/* Free the SPI port object */
+	if ((SPI *)(((struct mbed_spi_desc *)(desc->extra))->spi_port))
+		delete((SPI *)(((struct mbed_spi_desc *)(desc->extra))->spi_port));
+
+	/* Free the SPI extra descriptor object */
+	if ((struct mbed_spi_desc *)(desc->extra))
+		free((struct mbed_spi_desc *)(desc->extra));
+
+	/* Free the SPI descriptor object */
+	free(desc);
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Write and read data to/from SPI.
+ * @param desc[in] - The SPI descriptor.
+ * @param data[in, out] - The buffer with the transmitted/received data.
+ * @param bytes_number[in] - Number of bytes to write/read.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t mbed_spi_write_and_read(struct spi_desc *desc,
+				uint8_t *data,
+				uint16_t bytes_number)
+{
+	mbed::SPI *spi; 		// pointer to new spi instance
+	mbed::DigitalOut *csb;	// pointer to new CSB instance
+
+	if (!desc || !desc->extra || !data)
+		return -EINVAL;
+
+	spi = (SPI *)(((struct mbed_spi_desc *)(desc->extra))->spi_port);
+
+	if (((struct mbed_spi_desc *)desc->extra)->use_sw_csb) {
+		csb = (DigitalOut *)(((struct mbed_spi_desc *)(desc->extra))->csb_gpio);
+		csb->write(GPIO_LOW);
+	}
+
+	/* Perform synchronous SPI write and read */
+	spi->write((const char *)data, bytes_number, (char *)data, bytes_number);
+
+	if (((struct mbed_spi_desc *)desc->extra)->use_sw_csb) {
+		csb->write(GPIO_HIGH);
+	}
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Transfer (write/read) the number of SPI messages.
+ * @param desc[in] - The SPI descriptor.
+ * @param msgs[in, out] - Pointer to SPI messages.
+ * @param num_of_msgs[in] - Number of SPI messages.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ * @note Use of this function requires CSB pin to be software controlled.
+ */
+int32_t mbed_spi_transfer(struct spi_desc *desc, struct spi_msg *msgs,
+			  uint32_t num_of_msgs)
+{
+	mbed::SPI *spi; 			// pointer to new spi instance
+	mbed::DigitalOut *csb;		// pointer to new CSB instance
+	uint8_t msg_cnt;			// SPI message counter
+
+	if (!desc || !desc->extra || !msgs)
+		return -EINVAL;
+
+	if (!((struct mbed_spi_desc *)desc->extra)->use_sw_csb)
+		return -EINVAL;
+
+	spi = (SPI *)(((struct mbed_spi_desc *)(desc->extra))->spi_port);
+	csb = (DigitalOut *)(((struct mbed_spi_desc *)(desc->extra))->csb_gpio);
+
+	for (msg_cnt = 0; msg_cnt < num_of_msgs; msg_cnt++) {
+		csb->write(GPIO_LOW);
+
+		/* Perform synchronous SPI write and read */
+		if (!msgs[msg_cnt].tx_buff) {
+			spi->write(NULL, 0,
+				   (char *)msgs[msg_cnt].rx_buff, msgs[msg_cnt].bytes_number);
+		} else {
+			spi->write((const char *)msgs[msg_cnt].tx_buff, msgs[msg_cnt].bytes_number,
+				   (char *)msgs[msg_cnt].rx_buff, msgs[msg_cnt].bytes_number);
+		}
```

Comment:
Is this check necessary? If we use just the code in the else branch all possibilities should be covered, if I'm not mistaken.

Reply:
We need to pass transmit byte count as 0 and transmit buffer NULL if just intended to read the data from SPI. The function overloading is used for spi->write() function in Mbed layer.
Also the 2nd argument to this function i.e. transmit byte count is not explicitly passed to spi_transfer() function and hence some sort of condition is needed for setting transmit count as 0.